test/unit/math/test_ad.hpp

```diff
@@ -1446,15 +1475,51 @@ void expect_ad_vectorized_binary(const F& f, const T1& x, const T2& y) {
  * @param A_mv_f The result of a function from applying `A_mv`.
  * @param A_vm_f The result of a function from applying `A_vm`
  */
-template <typename MatVar, typename VarMat, typename ResultMatVar,
-          typename ResultVarMat, require_var_t<ResultMatVar>* = nullptr>
-inline void test_matvar_gradient(MatVar&& A_mv, VarMat&& A_vm,
-                                 ResultMatVar&& A_mv_f, ResultVarMat&& A_vm_f) {
+template <typename ResultMatVar, typename ResultVarMat, typename MatVar,
+          typename VarMat, require_var_t<ResultMatVar>* = nullptr>
+inline void test_matvar_gradient(ResultMatVar& A_mv_f, ResultVarMat& A_vm_f,
+                                 const MatVar& A_mv, const VarMat& A_vm) {
   A_vm_f.adj() = 1;
   A_mv_f.adj() = 1;
   stan::math::grad();
-  expect_near_rel("var<Matrix> vs Matrix<var> input adjoints", A_vm.adj(),
-                  A_mv.adj(), 1e-12);
+  expect_near_rel_var("var<Matrix> vs Matrix<var> result", A_vm, A_mv);
+  expect_near_rel("var<Matrix> vs Matrix<var> result value", A_vm_f.val(),
+                  A_mv_f.val(), 1e-12);
+  stan::math::set_zero_all_adjoints();
+}
+
+/**
+ * Test that the jacobian for matrices of vars is equal for the
+ *  var matrix when the result is an `var` type.
+ * @tparam MatVar1 An Eigen type inheriting from `EigenBase` that has
+ *  Scalar vars.
+ * @tparam MatVar2 An Eigen type inheriting from `EigenBase` that has
+ *  Scalar vars.
+ * @tparam VarMat1 A `var_value` with an inner type inheriting from `EigenBase`.
+ * @tparam VarMat2 A `var_value` with an inner type inheriting from `EigenBase`.
+ * @tparam ResultMatVar The resulting type of applying a function to `A_mv1`
+ * and `A_mv2`.
+ * @tparam ResultVarMat The result type of applying a function to `A_vm1` and
+ *  `A_vm2`.
+ * @param A_mv1 A eigen type of vars.
+ * @param A_mv2 A eigen type of vars.
+ * @param A_vm1 A eigen type of vars.
+ * @param A_vm2 A eigen type of vars.
+ * @param A_mv_f The result of a function from applying `A_mv1` and `A_mv2`.
+ * @param A_vm_f The result of a function from applying `A_vm1` and `A_vm2`
+ */
+template <typename ResultMatVar, typename ResultVarMat, typename MatVar1,
+          typename MatVar2, typename VarMat1, typename VarMat2,
+          require_all_var_vt<std::is_arithmetic, ResultMatVar,
+                             ResultVarMat>* = nullptr>
```

Comment:
What is this require logic saying here?

Reply:
This one just requires that the `ResultMatVar` and `ResultVarMat` are both `var_value<double>`. So it checks whether both are `var_value` and whether the inner types are arithmetic.